gnovm/pkg/gnolang/type_check.go

```diff
@@ -267,17 +267,12 @@ Main:
 				if fv, ok := cx.V.(*FuncValue); ok {
 					if fv.PkgPath == uversePkgPath {
 						// TODO: should support min, max, real, imag
-						switch {
-						case fv.Name == "len":
+						switch fv.Name {
+						case "len", "cap":
 							at := evalStaticTypeOf(store, last, currExpr.Args[0])
-							if _, ok := unwrapPointerType(baseOf(at)).(*ArrayType); ok {
-								// ok
-								break Main
+							if AsExpr(currExpr.Args[0], &CallExpr{}, &seen[Expr]{}) {
+								panic(fmt.Sprintf("%s (value of type %s) is not constant", currExpr.String(), at))
```

Comment:
can we just check into the compositLit's  Elts, and  do:
`assertValidConstValue(store, last, elt.Value)` ?

Reply:
Your point is that the composite literal expression I think can't be used here due to the rule:

> The expressions len(s) and cap(s) are constants if the type of s is an array or pointer to an array and the expression s does not contain channel receives or (non-constant) function calls; in this case s is not evaluated.

For instance, in the following code:

```go
package main

import "fmt"

func S() [2]int {
	return [2]int{}
}

func main() {
	const s = len(S())
	fmt.Println("Hello, 世界", s)
}
```

the call to `S()` is a function call, which makes `len(S())` non-constant.